crates/shared/src/remaining_amounts.rs

```diff
@@ -222,9 +270,109 @@ mod tests {
+        let order = Order {
+            data: OrderData {
+                sell_amount: 1000.into(),
+                buy_amount: 2000.into(),
+                fee_amount: 337.into(),
+                kind: OrderKind::Sell,
+                partially_fillable: false,
+                ..Default::default()
+            },
+            ..Default::default()
+        };
+        let remaining =
+            Remaining::from_order_with_balance(&order, order.data.sell_amount - 1).unwrap();
+        assert_eq!(remaining.remaining(1000.into()).unwrap(), 0.into());
+
+        // A partially fillable order that has not been executed at all scales
+        // to the available balance.
+        let order = Order {
+            data: OrderData {
+                sell_amount: 800.into(),
+                buy_amount: 2000.into(),
+                fee_amount: 200.into(),
+                kind: OrderKind::Sell,
+                partially_fillable: true,
+                ..Default::default()
+            },
+            ..Default::default()
+        };
+        {
+            // More than enough balance for the full order.
+            let remaining = Remaining::from_order_with_balance(&order, 5000.into()).unwrap();
+            assert_eq!(remaining.remaining(800.into()).unwrap(), 800.into());
+        }
+        {
+            // Not enough balance for the full order.
+            let remaining = Remaining::from_order_with_balance(&order, 500.into()).unwrap();
+            assert_eq!(remaining.remaining(800.into()).unwrap(), 400.into());
+        }
+
+        // A partially fillable order that has has been partially executed scales
+        // to the remaining execution and available balance.
+        let order = Order {
+            data: OrderData {
+                sell_amount: 800.into(),
+                buy_amount: 2000.into(),
+                fee_amount: 200.into(),
+                kind: OrderKind::Sell,
+                partially_fillable: true,
+                ..Default::default()
+            },
+            metadata: OrderMetadata {
+                executed_sell_amount_before_fees: 400.into(),
+                ..Default::default()
+            },
+            ..Default::default()
+        };
+        {
+            // More than enough balance for the full order.
+            let remaining = Remaining::from_order_with_balance(&order, 5000.into()).unwrap();
+            assert_eq!(remaining.remaining(800.into()).unwrap(), 400.into());
+        }
+        {
+            // Not enough balance for the full order.
+            let remaining = Remaining::from_order_with_balance(&order, 250.into()).unwrap();
+            assert_eq!(remaining.remaining(800.into()).unwrap(), 200.into());
+        }
+    }
+
+    #[test]
+    fn support_scaling_for_large_orders_with_partial_balance() {
+        let u = <U256 as From<u128>>::from;
+        let order = Order {
+            data: OrderData {
+                sell_amount: u(10).pow(u(30)),
```

Comment:
This way you don't need `u` at all.
```suggestion
                sell_amount: U256::exp10(30),
```